### PR TITLE
Remove CountUpAndDownLatch & refactor ConcurrentBattleCalculator

### PR DESCRIPTION
**Pull request by RoiEXLab**
## Change Summary & Additional Notes

This refactors the `ConcurrentBattleCalculator` to use a normal synchronized block instead of custom untested latch logic. The behaviour should be pretty similar from a thread safety and synchronisation standpoint (hopefully even a little bit better). I rand a full pro AI game of TWW 3 for a couple of rounds and no exception occured (except for #10820 which is unrelated, I'm getting it occasionally).

### CODE REVIEW

**Comment by RoiEXLab** [comment]
Hmm the smoke test failure (on builds 1 and 2 on this PR) are weird:

```
AiGameTest > testAiGame(String, String) > games.strategy.engine.data.AiGameTest.testAiGame(String, String)[1] FAILED
    java.lang.NullPointerException
        at games.strategy.triplea.odds.calculator.BattleCalculator.calculate(BattleCalculator.java:79)
        at games.strategy.triplea.odds.calculator.ConcurrentBattleCalculator.lambda$calculate$4(ConcurrentBattleCalculator.java:178)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:952)
        at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:926)
        at java.base/java.util.stream.AbstractTask.compute(AbstractTask.java:327)
        at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinTask.doInvoke(ForkJoinTask.java:408)
        at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:736)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateParallel(ReduceOps.java:919)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
        at games.strategy.triplea.odds.calculator.ConcurrentBattleCalculator.calculate(ConcurrentBattleCalculator.java:190)
        at games.strategy.triplea.ai.pro.util.ProOddsCalculator.callBattleCalc(ProOddsCalculator.java:250)
        at games.strategy.triplea.ai.pro.util.ProOddsCalculator.callBattleCalc(ProOddsCalculator.java:226)
        at games.strategy.triplea.ai.pro.util.ProOddsCalculator.calculateBattleResults(ProOddsCalculator.java:159)
        at games.strategy.triplea.ai.pro.util.ProOddsCalculator.calculateBattleResults(ProOddsCalculator.java:143)
        at games.strategy.triplea.ai.pro.ProCombatMoveAi.determineTerritoriesThatCanBeHeld(ProCombatMoveAi.java:498)
        at games.strategy.triplea.ai.pro.ProCombatMoveAi.doCombatMove(ProCombatMoveAi.java:92)
        at games.strategy.triplea.ai.pro.AbstractProAi.purchase(AbstractProAi.java:225)
        at games.strategy.triplea.ai.AbstractAi.start(AbstractAi.java:508)
        at games.strategy.engine.framework.ServerGame.waitForPlayerToFinishStep(ServerGame.java:593)
        at games.strategy.engine.framework.ServerGame.runStep(ServerGame.java:450)
        at games.strategy.engine.framework.ServerGame.runNextStep(ServerGame.java:345)
        at games.strategy.engine.data.AiGameTest.testAiGame(AiGameTest.java:44)
```

It seems like there was a NPE because `gameData` was null, but because the variable is final and is guaranteed to be non-null when invoking the constructor this exception should never happen at all. It's fixed now, but odd nevertheless

File: game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/ConcurrentBattleCalculator.java [hunk 1]
```diff
@@ -154,25 +148,11 @@ private void createWorkers(final GameData data) {
     if (cancelCurrentOperation.get() < 0 || data == null) {
       // we could have cancelled while setting data, so clear the workers again if so
       workers.clear();
-      isDataSet = false;
+      return false;
     } else {
```
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/ConcurrentBattleCalculator.java:152]: Agreed

File: game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/ConcurrentBattleCalculator.java [hunk 2]
```diff
@@ -33,64 +32,59 @@ public class ConcurrentBattleCalculator implements IBattleCalculator {
...
+    } catch (final InterruptedException e) {
+      Thread.currentThread().interrupt();
+    } catch (ExecutionException e) {
+      throw new IllegalStateException("CompletableFuture should handle all exceptions", e);
+    }
+  }
+
+  private void waitForGameDataReady() {
+    synchronized (mutexSetGameData) {
+      waitForGameDataReadyInternal();
```
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/ConcurrentBattleCalculator.java]: You're right it doesn't really hurt to have 2 synchronized blocks there.

I'm like 99% sure the current behaviour is mandatory to ensure correct operation. There are actually 2 mutexes involved here (I think you missed one).

Assuming waitForGameDataReady() is called from within the mutexCalcIsRunning block consider this example:

Example:
Thread 1 calls setGameData. The method is run until the end, but the future isn't started yet.
Thread 2 calls calculate. The synchronized block (mutexCalcIsRunning) is entered and waiting for the future to complete.
Thread 3 (the future) is starting and tries to acquire the mutexCalcIsRunning mutex to be able to start, which is currently held by thread 2 which waits for this thread to exit the mutex. This results in a deadlock!

[the remark this answers is not in the document]
Comment (asvitkine) [reply game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/ConcurrentBattleCalculator.java]: You're right, I did miss that they're separate mutexes.

Do we really need both or could just change to using a single one in both places without a big impact?

[the remark this answers is not in the document]
Comment (RoiEXLab) [reply game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/ConcurrentBattleCalculator.java]: Not sure. I try to be very cautious about the changes I do to this class. I tried way too often to change it causing a lot of weird race conditions to happen.
I think you might run into problems when trying to merge both mutexes, but I think those problems can be tackled.

File: game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/ConcurrentBattleCalculator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/ConcurrentBattleCalculator.java]: I'd prefer we keep the optional else block. It makes the flow easier to see from a block level, namely keeps the return statements on the same indention level which makes it a bit easier to grok.

**Pull request merged**